File: Residential_Controller.js
//MAIN
function main (number_of_floor, number_of_elevator){
    let elevator_controller = new ElevatorController(number_of_floor, number_of_elevator);
    elevator_controller.elevator_list[0].currentFloor = 3;
    elevator_controller.elevator_list[0].direction = null;
    elevator_controller.elevator_list[0].status = "IDLE";
    elevator_controller.elevator_list[0].floorList = [];

    elevator_controller.elevator_list[1].currentFloor = 10;
    elevator_controller.elevator_list[1].direction = null;
    elevator_controller.elevator_list[1].status = "IDLE";
    elevator_controller.elevator_list[1].floorList = [];
    elevator_controller.requestElevator(1,"UP");
    elevator_controller.RequestFloor(1,6);
    elevator_controller.requestElevator(3,"UP");
    elevator_controller.RequestFloor(1,5);
    elevator_controller.requestElevator(9,"DOWN");
    elevator_controller.RequestFloor(1,2);
    
}
class Elevator {
    constructor(number_of_elevator, number_of_floor){
        this.elevatorNumber = number_of_elevator;
        this.currentFloor = 1;
        this.direction = null;
        this.status = "idle";
        this.floorList = [];
        this.internalButton_list = [];
        for (let j = 1; j <= number_of_floor; j++){
            this.internalButton_list.push(new InternalButton(number_of_elevator, "off", j));
        }
       
        
    }   
}
class Button {
    constructor(direction, floor, status) {
        this.floor = floor;
        this.direction = direction;
        this.status = status;
    }
 }
 class InternalButton {
    constructor(number_of_elevator, status, floor){
    this.elevatorNumber = number_of_elevator;
    this.status = status;
    this.floor = floor;
    }
}
class ElevatorController {
    
    constructor(number_of_floor, number_of_elevator) {
 
        this.number_of_floor = number_of_floor;
        this.number_of_elevator = number_of_elevator;
        this.button_list = [];
        this.elevator_list = [];
            for (let i = 1; i < number_of_floor; i++) {
                this.button_list.push(new Button('UP', i, 'off'));
                this.button_list.push(new Button('DOWN', i + 1, 'off'));
            }
            for (let i = 1; i <= number_of_elevator; i++ ){
                this.elevator_list.push(new Elevator(i,number_of_floor));
            } 
            
    }
    //REQUEST ELEVATOR
    requestElevator(floorNumber, direction) {             
      
     
         var Elevator = this.findElevator(floorNumber, direction, this.elevator_list)
        
         console.log("User is at floor ", floorNumber, "and is going ", direction);
         console.log("Elevator responding to the request:")
         console.log(Elevator);  
         
         this.bubbleSort(Elevator.floorList, Elevator.direction)
         
         this.operateElevator(Elevator, direction)
        
     }
        
     /*
    //ACTIVATE LIGHT
    light (floorNumber, direction, list) {    
        list.forEach(function(button)  {
                console.log(button)
            if ( (floorNumber == button.floor) && (direction == button.direction) ) {
                button.status = "on"
                console.log("Bouton activate:")
                console.log(button)
            };   
        });
     console.log (list)
    }
    */
    //FIND ELEVATOR
    findElevator (floorNumber, direction, elevlist) {
        
       
            for(var i = 0; i < elevlist.length; i++){
                var e = elevlist[i];
                if(e.status === "STOPPED" && e.currentFloor === floorNumber && e.direction === direction){
                    e.floorList.push(floorNumber);
                    
                    return e;
            
                }else if(e.status === "IDLE" && e.currentFloor === floorNumber){
                    e.floorList.push(floorNumber);
                    
                    return e;
    
                }else if(e.currentFloor < floorNumber && (e.status === "MOVING" || "STOPPED") && e.direction === "UP" && direction === e.direction){  
                    e.floorList.push(floorNumber);
                    
                    return e;
    
                }else if(e.currentFloor > floorNumber && (e.status === "MOVING" || "STOPPED") && e.direction === "DOWN" && direction === e.direction){
                    e.floorList.push(floorNumber);
                    
                    return e;
                }else if(e.status === "IDLE"){
                    e.floorList.push(floorNumber);
                   
                    return e;
                }
                /*else if (i+1 === elevlist.length){
                    console.log("toto")
                    e.floorList.push(floorNumber);
                    return e;
                }*/
            }  
        
    }
    /*shortestList(elevlist){
        var length = 9999
       
        for(var i = 0; i < elevlist.length; i++){
            if( length > elevlist[i].floorList.length){
                length = elevlist[i].floorList.length
                var r = elevlist[i]
            }
        }
        return r;
    }*/
   
    //BUBBLESORT
    bubbleSort(floorList, direction){ 
        if (direction === "UP"){
            var swapped;
            do {
                swapped = false;
                for (var i=0; i < floorList.length-1; i++) {
                    if (floorList[i] > floorList[i+1]) {
                        var temp = floorList[i];
                        floorList[i] = floorList[i+1];
                        floorList[i+1] = temp;
                        swapped = true;
                    }
                }
            } while (swapped);
        }
        //DÉCROISSANT
        else if(direction === "DOWN"){
            do {
                swapped = false;
                for (var i=0; i < floorList.length-1; i++) {
                    if (floorList[i] < floorList[i+1]) {
                        var temp = floorList[i];
                        floorList[i] = floorList[i+1];
                        floorList[i+1] = temp;
                        swapped = true;
                    }
                }
            } while (swapped);
        }
        console.log("Sorted floorList : ", floorList);
        
    }
    
    operateElevator(Elevator, direction){
            console.log("next item on floor list : ", Elevator.floorList[0])
            
        while (Elevator.floorList.length > 0){
                console.log("current floor : ", Elevator.currentFloor)
            if (Elevator.floorList[0] === Elevator.currentFloor){
                this.openDoor(Elevator, direction);
                Elevator.floorList.shift()
                console.log("new list : ",Elevator.floorList);
                console.log("next destination : ", Elevator.floorList[0])
            }
            if (Elevator.floorList[0] > Elevator.currentFloor){
                this.moveUp(Elevator);
            }
            if (Elevator.floorList[0] < Elevator.currentFloor){
                this.moveDown(Elevator); 
            }
            
            
        }
            if (Elevator.floorList.length < 1){
                Elevator.status = "IDLE";
                Elevator.direction = null;
                console.log(Elevator);
            }
    }
    
    
    moveUp(Elevator){
        if(Elevator.status =="IDLE" || Elevator.status =="STOPPED"){
            Elevator.status = "MOVING";
            Elevator.direction = "UP";
            console.log("Elevator start ", Elevator.status, " ", Elevator.direction);
        }
            Elevator.currentFloor++;
        
    }
    moveDown(Elevator){
        if(Elevator.status =="IDLE" || Elevator.status =="STOPPED"){
        Elevator.status = "MOVING";
        Elevator.direction = "DOWN";
        console.log("elevator start moving", Elevator);
        }
        this.timer(2000);
        Elevator.currentFloor--;
       
    }
    openDoor(Elevator, direction){
        if( Elevator.status === "IDLE"){
            Elevator.direction = direction;
        }
        Elevator.status = "STOPPED";
        
        console.log("Elevator is ",Elevator.status);
        //TIMER
        this.timer(2000);
        console.log("DOOR OPENING");
        this.timer(2000);
        console.log("DOOR IS OPEN");
        this.timer(2000);
        this.closeDoor(Elevator);
    }

    closeDoor(Elevator){
    console.log("DOOR IS CLOSING");
    this.timer(2000);
    console.log("DOOR IS CLOSE");
    }

    timer(milliseconds) {
        var start = new Date().getTime();
        for (var i = 0; i < 1e7; i++) {
            if ((new Date().getTime() - start) > milliseconds) {
                break;
            }
        }
        
    };
            
        
                
    RequestFloor(Elevator, RequestedFloor) {
        console.log("User is in elevator number ", Elevator, " and is going to floor ", RequestedFloor);
        var Elevator =  this.interior_floorList(Elevator, RequestedFloor,this.elevator_list);
    
        this.bubbleSort(Elevator.floorList, Elevator.direction);
       
        this.operateElevator(Elevator, RequestedFloor);           
    }
        interior_floorList(Elevator, RequestedFloor, elevatorList){
            for(var i = 0; i < elevatorList.length; i++){
                if( Elevator === elevatorList[i].elevatorNumber){
                    elevatorList[i].floorList.push(RequestedFloor);
                    return elevatorList[i];
                }
            }          
        }
}
  
 
main(10,2)

File: Residential_Controller.py
#MAIN

import time

def main (number_of_floor, number_of_elevator):
    elevator_controller = ElevatorController(number_of_floor, number_of_elevator)
    elevator_controller.elevator_list[0].currentFloor = 3
    elevator_controller.elevator_list[0].direction = None
    elevator_controller.elevator_list[0].status = "IDLE"
    elevator_controller.elevator_list[0].floorList = []

    elevator_controller.elevator_list[1].currentFloor = 10
    elevator_controller.elevator_list[1].direction = None
    elevator_controller.elevator_list[1].status = "IDLE"
    elevator_controller.elevator_list[1].floorList = []

    #HELLO, PLEASE ENTER VALID DATA TO OPERATE THE ELEVATOR
    elevator_controller.requestElevator(1,"UP")
    elevator_controller.RequestFloor(1,6)

    elevator_controller.requestElevator(3,"UP")
    elevator_controller.RequestFloor(1,5)

    elevator_controller.requestElevator(9,"DOWN")
    elevator_controller.RequestFloor(1,2)

class Elevator:
    def __init__(self, number_of_elevator, number_of_floor):
        self.elevatorNumber = number_of_elevator
        self.currentFloor = 1
        self.direction = None
        self.status = "idle"
        self.floorList = []
        self.internalButton_list = []
        for j in range(number_of_floor): #(let j = 1; j <= number_of_floor; j++){
            self.internalButton_list.append(InternalButton(number_of_elevator, "off", j))
        
       
class Button:
    def __init__(self, direction, floor, status):
        self.floor = floor
        self.direction = direction
        self.status = status
    
 
class InternalButton:
    def __init__(self, number_of_elevator, status, floor):
        self.elevatorNumber = number_of_elevator
        self.status = status
        self.floor = floor
    

class ElevatorController:
    
    def __init__(self, number_of_floor, number_of_elevator):
        self.number_of_floor = number_of_floor
        self.number_of_elevator = number_of_elevator
        self.button_list = []
        self.elevator_list = []
        for i in range(number_of_floor):
            self.button_list.append(Button("UP",i,number_of_floor))
            self.button_list.append(Button("DOWN",i+1,number_of_floor))
        for i in range(number_of_elevator):
            self.elevator_list.append(Elevator(i,number_of_floor))
        
    #REQUEST ELEVATOR
    def requestElevator(self, floorNumber, direction):           
        Elevator = self.findElevator(floorNumber, direction, self.elevator_list)
        
        print("User is at floor ", floorNumber, "and wants to go ", direction)
        print("Elevator responding to the request: ", Elevator.elevatorNumber+1)
        

        self.bubbleSort(Elevator.floorList, Elevator.direction)
         
        self.operateElevator(Elevator, direction)
        
     
        
     
    #ACTIVATE LIGHT
   # light (floorNumber, direction, list) {    
    #    list.forEach(function(button)  {
        #        console.log(button)
          #  if ( (floorNumber == button.floor) && (direction == button.direction) ) {
           #     button.status = "on"
            #    console.log("Bouton activate:")
            #    console.log(button)
          #  };   
       # });
     #console.log (list)
    
    
    #FIND ELEVATOR
    def findElevator (self, floorNumber, direction, elevlist):
        for i in range(len(elevlist)):
            e = elevlist[i]
            
            if e.status == "STOPPED" and e.currentFloor == floorNumber and e.direction == direction:
                e.floorList.append(floorNumber)
                return e
            
            elif e.status == "IDLE" and e.currentFloor == floorNumber:
                e.floorList.append(floorNumber)   
                return e
    
            elif e.currentFloor < floorNumber and (e.status == "MOVING" or e.status == "STOPPED") and e.direction == "UP" and direction == e.direction: 
                e.floorList.append(floorNumber)
                return e
    
            elif e.currentFloor > floorNumber and (e.status == "MOVING" or e.status == "STOPPED") and e.direction == "DOWN" and direction == e.direction:
                e.floorList.append(floorNumber)
                return e

            elif e.status == "IDLE":
                e.floorList.append(floorNumber)
                return e
                
            #elif i+1 == elevlist.length:
              #  e.floorList.push(floorNumber)
              #  return e
            
             
        
    
    #shortestList(elevlist){
       # var length = 9999
       
        #for(var i = 0; i < elevlist.length; i++){
         #   if( length > elevlist[i].floorList.length){
         #       length = elevlist[i].floorList.length
         #       var r = elevlist[i]
          #  }
        #}
      #  return r;
   # }*/
   
    #BUBBLESORT
    #CRESCENT
    def bubbleSort(self, floorList, direction):
        if direction == "UP":
            for passnum in range(len(floorList)-1,0,-1):
                for i in range(passnum):
                    if floorList[i]>floorList[i+1]:
                        temp = floorList[i]
                        floorList[i] = floorList[i+1]
                        floorList[i+1] = temp
    #DESCENDING
    
        elif direction == "DOWN":
            for passnum in range(len(floorList)-1,0,-1):
                for i in range(passnum):
                    if floorList[i]<floorList[i+1]:
                        temp = floorList[i]
                        floorList[i] = floorList[i+1]
                        floorList[i+1] = temp
        print("Sorted floorList : ", floorList)

    #OPERATE ELEVATOR    
    def operateElevator(self, Elevator, direction):
        
        print("next item on floor list : ", Elevator.floorList[0])
            
        while (len(Elevator.floorList) > 0):
            
            print("current floor : ", Elevator.currentFloor)
            if (Elevator.floorList[0] == Elevator.currentFloor):
                self.openDoor(Elevator, direction)
                Elevator.floorList.pop(0)
                print("new list : ",Elevator.floorList)
                #print("next destination : ", Elevator.floorList[0])
            
            elif Elevator.floorList[0] > Elevator.currentFloor:
                self.moveUp(Elevator)
            
            elif Elevator.floorList[0] < Elevator.currentFloor:
                self.moveDown(Elevator)
            
        if (len(Elevator.floorList) < 1):
                Elevator.status = "IDLE"
                Elevator.direction = None
                print("Elevator is ", Elevator.status)
            
    
    
    #MOVE UP
    def moveUp(self, Elevator):
        if Elevator.status =="IDLE" or Elevator.status =="STOPPED":
            Elevator.status = "MOVING"
            Elevator.direction = "UP"
            print("elevator start moving", Elevator.direction)
        self.timer(2)
        Elevator.currentFloor = Elevator.currentFloor+1
        
    
    def moveDown(self, Elevator):
        if Elevator.status =="IDLE" or Elevator.status =="STOPPED":
            Elevator.status = "MOVING"
            Elevator.direction = "DOWN"
            print("elevator start moving", Elevator.direction)
        self.timer(2)
        Elevator.currentFloor = Elevator.currentFloor-1
       
    
    def openDoor(self, Elevator, direction):
        if( Elevator.status == "IDLE"):
            Elevator.direction = direction  
            Elevator.status = "STOPPED"
        
            print("Elevator is ",Elevator.status)

    #TIMER
        self.timer(2)
        print("DOOR OPENING")
        self.timer(2)
        print("DOOR IS OPEN")
        self.timer(2)
        self.closeDoor(Elevator)
    

    def closeDoor(self, Elevator):
        print("DOOR IS CLOSING")
        self.timer(2)
        print("DOOR IS CLOSE")
    

    def timer(self, milliseconds):
       time.sleep(milliseconds)
            
        
                
    def RequestFloor(self, Elevator, RequestedFloor):
        Elevator = Elevator -1
        print("User is in elevator number ", Elevator+1, " and wants to go floor ", RequestedFloor)
        Elevator =  self.interior_floorList(Elevator, RequestedFloor,self.elevator_list)
    
        self.bubbleSort(Elevator.floorList, Elevator.direction)
       
        self.operateElevator(Elevator, RequestedFloor)          
    
    def interior_floorList(self, Elevator, RequestedFloor, elevatorList):
        for i in range(len(elevatorList)):
            if Elevator == elevatorList[i].elevatorNumber:
                elevatorList[i].floorList.append(RequestedFloor)
                return elevatorList[i]
               
               

  
 
main(10,2)